File: Design-Patterns-Using-Kotlin/ConcurrencyPattern/DoubleChecking.kt
class Database private constructor() {
    init {
        println("Database created")
    }

    fun addData(data: String) {
        println(data)
    }

    companion object {
        @Volatile
        private var instance: Database? = null

        fun getInstance(): Database {
            if (instance == null) {
                synchronized(this) {
                    if (instance == null) {
                        instance = Database()
                    }
                }
            }
            return instance!!
        }
    }
}

// Client code.
fun main() {
    val s1 = Database.getInstance()
    val s2 = Database.getInstance()

    println(s1)
    println(s2)

    s2.addData("Hello, world!")
}

/*
Database created
Database@6bc7c054
Database@6bc7c054
Hello, world!
*/

File: Design-Patterns-Using-CSharp/ConcurrencyPattern/DoubleChecking.cs
using System;

public class Database
{
    public Database()
    {
        Console.WriteLine("Database created");
    }

    public void AddData(string data)
    {
        Console.WriteLine(data);
    }
}

public sealed class Singleton
{
    private static volatile Singleton instance;
    private static Database db;
    private static readonly object lockObject = new object();

    private Singleton()
    {
        db = new Database();
    }

    public static Singleton Instance
    {
        get
        {
            if (instance == null)
            {
                lock (lockObject)
                {
                    if (instance == null)
                    {
                        instance = new Singleton();
                    }
                }
            }
            return instance;
        }
    }

    public void AddData(string data)
    {
        db.AddData(data);
    }
}

// Client code

public class DoubleChecking
{
    public static void Main()
    {
        Singleton s1 = Singleton.Instance;
        Singleton s2 = Singleton.Instance;

        Console.WriteLine(s1);
        Console.WriteLine(s2);

        s2.AddData("Hello, world!");
    }
}

/*
Database created
Singleton
Singleton
Hello, world!
*/